src/sysupdate/updatectl.c

```diff
@@ -0,0 +1,1219 @@
+                const char *class, *name, *path;
+
+                r = bus_call_method(bus, bus_sysupdate_mgr, "ListTargets", &error, &reply, NULL);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to call ListTargets: %s", bus_error_message(&error, r));
+
+                r = sd_bus_message_enter_container(reply, 'a', "(sso)");
+                if (r < 0)
+                        return r;
+
+                while ((r = sd_bus_message_read(reply, "(sso)", &class, &name, &path)) > 0) {
+                        _cleanup_free_ char *id = NULL;
+
+                        if (streq(class, "host"))
+                                id = strdup("host");
+                        else
+                                id = strjoin(class, ":", name);
+                        if (!id)
+                                return log_oom();
+
+                        r = strv_consume(&targets, TAKE_PTR(id));
+                        if (r < 0)
+                                return r;
+                }
+                if (r < 0)
+                        return r;
+
+                r = sd_bus_message_exit_container(reply);
+                if (r < 0)
+                        return r;
+
+        } else {
+                r = strv_extend_strv(&targets, argv, true);
+                if (r < 0)
+                        return r;
+        }
+
+        *ret_targets = TAKE_PTR(targets);
+        return 0;
+}
+
+static int parse_target(
+                const char *in,
+                char **ret_bus_path,
+                char **ret_version) {
+        _cleanup_free_ char *id = NULL, *version = NULL;
+        _cleanup_free_ char *escaped = NULL, *objpath = NULL;
+        char *s;
+
+        /*
+         * Parses the TARGET[@VERSION] syntax from the command line into
+         * a bus object locator and an optional version number.
+         */
+
+        assert(in);
+        assert(ret_bus_path);
+        assert(ret_version);
+
+        s = strrchr(in, '@');
+        if (s) {
+                version = strdup(s + 1);
+                if (!version)
+                        return -ENOMEM;
+                id = strndup(in, s - in);
+        } else
+                id = strdup(in);
+        if (!id)
+                return -ENOMEM;
+
+        escaped = bus_label_escape(id);
+        if (!escaped)
+                return -ENOMEM;
+
+        objpath = strjoin("/org/freedesktop/sysupdate1/target/", escaped);
+        if (!objpath)
+                return -ENOMEM;
+
+        *ret_bus_path = TAKE_PTR(objpath);
+        *ret_version = TAKE_PTR(version);
+        return 0;
+}
+
+static int parse_targets(
+                char **targets,
+                size_t *ret_n,
+                char ***ret_bus_paths,
+                char ***ret_versions) {
+        _cleanup_strv_free_ char **bus_paths = NULL;
+        _cleanup_strv_free_ char **versions = NULL;
+        size_t n = 0;
+        int r;
+
+        assert(targets);
```

Comment:
This crashes whenever I try `updatectl` (on a non-sysupdate distro).

A quick look and it seems this case is better handled by the `strv_isempty` check below, which would return a clearer error.

Reply:
Good catch, removed the `assert` to just go straight to the `strv_isempty` check.